pkg/fs/fs_unix.go

```diff
@@ -40,7 +40,11 @@ func SyncDir(dirName string) error {
 // of oldpath. If this function returns successfully, the contents of newpath will
 // be identical to oldpath, and oldpath will be removed.
 func RenameFileWithReplacement(oldpath, newpath string) error {
-	return os.Rename(oldpath, newpath)
+	if err := os.Rename(oldpath, newpath); err == nil {
+		return nil
+	}
+
+	return MoveFileWithReplacement(oldpath, newpath)
```

Comment:
The `MoveFileWithReplacement` call is made no matter what the error returned from `Rename`. It would be preferable only to try the copy if the error indicates that the two file locations are on different file systems (as opposed to the destination being nonexistent, or the source, or some other error)  Alternately, you could test with [syscall.Fstatfs](https://pkg.go.dev/syscall#Fstatfs) to determine if the file move spans two file systems.

Reply:
True. I changed it to using ``syscall.Stat()`` and check the ``dev_t`` coming from that call.

Reply:
Okay, we might have a problem here. I realized that ``Stat()`` alone doesn't cut it, since the destination file usually does not exist, so I can't use ``Stat()`` here. ``Statfs()`` won't help either because (kind of surprisingly) it does not contain the device id, but only a thing called ``f_fsid``. ``man statfs`` has to say this about this field:

```
Nobody knows what f_fsid is supposed to contain (but see below).
```

In summary, it's quite a pain to ask for the right device beforehand. The old lazy approach had this easily covered and was more efficient surprisingly since we didn't have to several syscalls beforehand.

Reply:
As I wrote [here](https://github.com/influxdata/influxdb/pull/22909#discussion_r759760302) you may be able to check the  error return from `Rename` without adding system calls.  The rename should fail with `EXDEV` if the two link points are on different mounted filesystems (although it is important to verify this, of course).

>   [EXDEV](https://man7.org/linux/man-pages/man2/link.2.html)  oldpath and newpath are not on the same mounted
              filesystem.  (Linux permits a filesystem to be mounted at
              multiple points, but link() does not work across different
              mounts, even if the same filesystem is mounted on both.)

We don't want to try copying if the error is not caused by differing file system mount points.

Reply:
Right, totally forgot about that option. Implemented it that way as of 700d938c53d14cd30163cb29b5e0b7fb360ad63a.